### ISSUE TITLE
Issue with DH Group Exchange Request

### ISSUE THREAD

**Issue by ClemGit** [phpseclib/phpseclib#2035]
Hello

We are having recurring issues that seem more prevalent with the latest v2 version (2.0.47 while we were on 2.0.37 before) during the Diffie-Hellman Group Exchange Request.

The issues are random and it is difficult to identify the reasons behind their occurrence.
Unfortunately, I don't have library logs but I can see via tcpdump what appears to be an invalid behavior when constructing the request in some cases.

**Working case:**
The DH Group Exchange Request is 90 bytes in size
![image](https://github.com/user-attachments/assets/1bda9bbc-7411-4cb1-9c32-aa0d68bcd533)
And contains a standard payload
![image](https://github.com/user-attachments/assets/4898ffc1-f6c4-4bd4-bc01-0dc12c50959a)

**Non-working case:**
The DH Group Exchange Request is 142 bytes in size
![image](https://github.com/user-attachments/assets/765bc5f1-f2be-497d-af5a-99ba24863cb1)
And contains a payload with the client identifier
![image](https://github.com/user-attachments/assets/73e31809-68e5-4530-866a-c22170c5990b)
Which is normally sent at the onset of the connection
![image](https://github.com/user-attachments/assets/159c0fd4-df38-4f71-9cea-591add30de56)
The server then considers the message invalid as the packet size exceeds the maximum allowed size
![image](https://github.com/user-attachments/assets/16683d5b-8283-4f20-bf37-8cd40ab27814)

From our historical traces, this may have been happening with version 2.0.37 as well but it is difficult to assess.
We are running with PHP 8.1.

Are there known issues in version 2 that may be solved in version 3?
Is this is a new issue?

Thank you

**Comment by terrafrost** [comment]
I recognize the use of Wireshark from the screenshots!

Anyway,

> The DH Group Exchange Request is 142 bytes in size
> ![image](https://private-user-images.githubusercontent.com/2393191/360476466-765bc5f1-f2be-497d-af5a-99ba24863cb1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ1OTI1NjAsIm5iZiI6MTcyNDU5MjI2MCwicGF0aCI6Ii8yMzkzMTkxLzM2MDQ3NjQ2Ni03NjViYzVmMS1mMmJlLTQ5N2QtYWY1YS05OWJhMjQ4NjNjYjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjVUMTMyNDIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzNlY2M0N2RjZmM5OGI0NjUyYjNmMTAyY2M3YTBhZWU5NjEyYzMzZjQxNTk3YmIxZDhhOWI1YzJmZjc1NTQwNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.VroOnKJLJUSVlCxLRLGKMEHeYJKOxFapg444b47rY98)
> And contains a payload with the client identifier
> ![image](https://private-user-images.githubusercontent.com/2393191/360477078-73e31809-68e5-4530-866a-c22170c5990b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ1OTI1NjAsIm5iZiI6MTcyNDU5MjI2MCwicGF0aCI6Ii8yMzkzMTkxLzM2MDQ3NzA3OC03M2UzMTgwOS02OGU1LTQ1MzAtODY2YS1jMjIxNzBjNTk5MGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjVUMTMyNDIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2NkODQ5YzAwMWYzYjUwMWU4MDc3YjAyOWI0MzQyNTYyNmVlN2Q0YWU0NjM4ODkwYzBiMzg1ZGVmOTE1ZDM1YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.dvf3Z7Exho-VJy-ao3Ppmlxs_8H0zA3YZFgAZ45wU0I)

I have no idea how that could happen. From the phpseclib source code:

```php
        if (strpos($this->kex_algorithm, 'curve25519-sha256') === 0 || strpos($this->kex_algorithm, 'ecdh-sha2-nistp') === 0) {
            $curve = strpos($this->kex_algorithm, 'curve25519-sha256') === 0 ?
                'Curve25519' :
                substr($this->kex_algorithm, 10);
            $ourPrivate = EC::createKey($curve);
            $ourPublicBytes = $ourPrivate->getPublicKey()->getEncodedCoordinates();
            $clientKexInitMessage = 'NET_SSH2_MSG_KEX_ECDH_INIT';
            $serverKexReplyMessage = 'NET_SSH2_MSG_KEX_ECDH_REPLY';
        } else {
            if (strpos($this->kex_algorithm, 'diffie-hellman-group-exchange') === 0) {
                $dh_group_sizes_packed = pack(
                    'NNN',
                    $this->kex_dh_group_size_min,
                    $this->kex_dh_group_size_preferred,
                    $this->kex_dh_group_size_max
                );
                $packet = pack(
                    'Ca*',
                    NET_SSH2_MSG_KEXDH_GEX_REQUEST,
                    $dh_group_sizes_packed
                );
                $this->send_binary_packet($packet);
                $this->updateLogHistory('UNKNOWN (34)', 'NET_SSH2_MSG_KEXDH_GEX_REQUEST');

                $response = $this->get_binary_packet_or_close(NET_SSH2_MSG_KEXDH_GEX_GROUP);
                list($type, $primeBytes, $gBytes) = Strings::unpackSSH2('Css', $response);
                $this->updateLogHistory('NET_SSH2_MSG_KEXDH_REPLY', 'NET_SSH2_MSG_KEXDH_GEX_GROUP');
                $prime = new BigInteger($primeBytes, -256);
                $g = new BigInteger($gBytes, -256);

                $exchange_hash_rfc4419 = $dh_group_sizes_packed . Strings::packSSH2(
                    'ss',
                    $primeBytes,
                    $gBytes
                );

                $params = DH::createParameters($prime, $g);
                $clientKexInitMessage = 'NET_SSH2_MSG_KEXDH_GEX_INIT';
                $serverKexReplyMessage = 'NET_SSH2_MSG_KEXDH_GEX_REPLY';
            } else {
                $params = DH::createParameters($this->kex_algorithm);
                $clientKexInitMessage = 'NET_SSH2_MSG_KEXDH_INIT';
                $serverKexReplyMessage = 'NET_SSH2_MSG_KEXDH_REPLY';
            }

            $keyLength = min($kexHash->getLengthInBytes(), max($encryptKeyLength, $decryptKeyLength));

            $ourPrivate = DH::createKey($params, 16 * $keyLength); // 2 * 8 * $keyLength
            $ourPublic = $ourPrivate->getPublicKey()->toBigInteger();
            $ourPublicBytes = $ourPublic->toBytes(true);
        }

        $data = pack('CNa*', constant($clientKexInitMessage), strlen($ourPublicBytes), $ourPublicBytes);

        $this->send_binary_packet($data);
```
`$this->send_binary_packet()` is called twice, in succession, and the second time it's being sent it's sending a NET_SSH2_MSG_KEXDH_GEX_INIT packet - not an SSH identifier. In theory, calling `$this->get_binary_packet_or_close()` can result in SSH packets being sent but that'd only send NET_SSH2_MSG_DISCONNECT packets:

```php
    private function get_binary_packet_or_close(...$message_types)
    {
        try {
            $packet = $this->get_binary_packet();
            if (count($message_types) > 0 && !in_array(ord($packet[0]), $message_types)) {
                $this->disconnect_helper(NET_SSH2_DISCONNECT_PROTOCOL_ERROR);
                throw new ConnectionClosedException('Bad message type. Expected: #'
                    . implode(', #', $message_types) . '. Got: #' . ord($packet[0]));
            }
            return $packet;
        } catch (TimeoutException $e) {
            $this->disconnect_helper(NET_SSH2_DISCONNECT_BY_APPLICATION);
            throw new ConnectionClosedException('Connection closed due to timeout');
        }
    }

    protected function disconnect_helper($reason)
    {
        if ($this->bitmap & self::MASK_CONNECTED) {
            $data = Strings::packSSH2('CNss', NET_SSH2_MSG_DISCONNECT, $reason, '', '');
            try {
                $this->send_binary_packet($data);
            } catch (\Exception $e) {
            }
        }

        $this->reset_connection();

        return false;
    }
```
`$this->get_binary_packet_or_close()` also calls `$this->filter()`. Here's the source code for the first part of that function:

```php
        switch (ord($payload[0])) {
            case NET_SSH2_MSG_DISCONNECT:
                Strings::shift($payload, 1);
                list($reason_code, $message) = Strings::unpackSSH2('Ns', $payload);
                $this->errors[] = 'SSH_MSG_DISCONNECT: ' . self::$disconnect_reasons[$reason_code] . "\r\n$message";
                $this->disconnect_helper(NET_SSH2_DISCONNECT_CONNECTION_LOST);
                throw new ConnectionClosedException('Connection closed by server');
            case NET_SSH2_MSG_IGNORE:
                $this->extra_packets++;
                $payload = $this->get_binary_packet();
                break;
            case NET_SSH2_MSG_DEBUG:
                $this->extra_packets++;
                Strings::shift($payload, 2); // second byte is "always_display"
                list($message) = Strings::unpackSSH2('s', $payload);
                $this->errors[] = "SSH_MSG_DEBUG: $message";
                $payload = $this->get_binary_packet();
                break;
            case NET_SSH2_MSG_UNIMPLEMENTED:
                break; // return payload
            case NET_SSH2_MSG_KEXINIT:
                // this is here for key re-exchanges after the initial key exchange
                if ($this->session_id !== false) {
                    if (!$this->key_exchange($payload)) {
                        $this->disconnect_helper(NET_SSH2_DISCONNECT_KEY_EXCHANGE_FAILED);
                        throw new ConnectionClosedException('Key exchange failed');
                    }
                    $payload = $this->get_binary_packet();
                }
                break;
            case NET_SSH2_MSG_EXT_INFO:
                Strings::shift($payload, 1);
                list($nr_extensions) = Strings::unpackSSH2('N', $payload);
                for ($i = 0; $i < $nr_extensions; $i++) {
                    list($extension_name, $extension_value) = Strings::unpackSSH2('ss', $payload);
                    if ($extension_name == 'server-sig-algs') {
                        $this->supported_private_key_algorithms = explode(',', $extension_value);
                    }
                }
                $payload = $this->get_binary_packet();
        }
```
Most of those don't send data. NET_SSH2_MSG_DISCONNECT can but I've already discussed that situation. NET_SSH2_MSG_KEXINIT can iff `$this->session_id !== false` but is only set after NET_SSH2_MSG_KEXDH_GEX_REPLY is received:

```php
        if ($this->session_id === false) {
            $this->session_id = $this->exchange_hash;
        }
```
The only other possible explanation I can think of for what's going on is that PHP - or perhaps something even lower than that - is trying to "bundle" TCP/IP packets under certain conditions and it's messing up.

**Comment by terrafrost** [comment]
> Do you think it is worth moving to v3?

Yes.

> Is there anything I could do to help you investigate?

Maybe. You said the issue is intermittent. Well, technically, you said random, but same thing.

Anyway, how intermittent? Does it occur 10% of the time? 0.001% of the time? If the latter then trying to reproduce it with any regularity seems like it'd be quite difficult. And if it's happening that infrequently I wouldn't dismiss the possibility of [single event upsets](https://en.wikipedia.org/wiki/Single-event_upset) yielding unpredictable behavior.

And does it always happen with the same packets? Like is it always a NET_SSH2_MSG_KEXDH_GEX_INIT packet with the server identification string appended to it?

If it were happening on my machine and in lieu of knowing the answers to the above questions, one thing I might do, just as a sanity check, would be to add `echo "sending GEX INIT"` just above the line that actually sends the `NET_SSH2_MSG_KEXDH_GEX_INIT` packet and a `echo "sending server identification"` line just before that's actually sent. That, in theory, should provide empirical evidence that phpseclib is sending them in one order and that somehow, at some layer below PHP, are being re-ordered and concatenated together. The code change to do that would be as follows:

```php
#
#-----[ OPEN ]------------------------------------------
#
Net/SSH2.php
#
#-----[ FIND ]------------------------------------------
# this line occurs twice; this FIND is for the first occurrence
# it appears twice because of sendIdentificationStringFirst() / sendIdentificationStringLast(),
# which let you tell phpseclib to send it either first --or-- last
#
fputs($this->fsock, $this->identifier . "\r\n");
#
#-----[ BEFORE, ADD ]-----------------------------------
#
echo "sending server identification\n";
#
#-----[ FIND ]------------------------------------------
# this line occurs twice; this FIND is for the second occurrence
#
fputs($this->fsock, $this->identifier . "\r\n");
#
#-----[ BEFORE, ADD ]-----------------------------------
#
echo "sending server identification\n";
#
#-----[ FIND ]------------------------------------------
#
                $packet = pack(
                    'Ca*',
                    NET_SSH2_MSG_KEXDH_GEX_REQUEST,
                    $dh_group_sizes_packed
                );
                $this->send_binary_packet($packet);
#
#-----[ BEFORE, ADD ]-----------------------------------
#
echo "sending GEX INIT\n";
```

**Comment by ClemGit** [comment]
Happened around 20% of the time so I guess not an SEU.
Since we rolled back to V2.0.37 and PHP 7.4, no more issues.

One thing I just noticed in the wireshark traces is that when it works fine, the "Info" is "Client: Diffie-Hellman Group Exchange Request", while when it fails, it was just "Diffie-Hellman Group Exchange Request", without the "Client: ".

A working case:
![image](https://github.com/user-attachments/assets/5f1dbb27-195d-4c4e-8f9f-4be3f6536c64)

A non working case:
![Uploading image.png…]()

All packets before have the same size, that might rule out network optimization.

We'll give a try with v3 and adding traces and report back.
Thank you
